controllers/remote/cluster_cache.go

```diff
@@ -54,319 +48,206 @@ const (
+	}, nil
 }
 
-// getClusterCache returns the clusterCache for cluster, or nil if it does not exist.
-func (m *ClusterCacheTracker) getDelegatingClient(cluster client.ObjectKey) *client.DelegatingClient {
-	m.delegatingClientsLock.RLock()
-	defer m.delegatingClientsLock.RUnlock()
+// deleteAccessor stops an accessor's cache and removes the accessor from the tracker.
+func (t *ClusterCacheTracker) deleteAccessor(cluster client.ObjectKey) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
 
-	return m.delegatingClients[cluster]
-}
+	a, exists := t.accessors[cluster]
+	if !exists {
+		return
+	}
 
-// newDelegatingClient creates a new delegating client.
-func (m *ClusterCacheTracker) newDelegatingClient(ctx context.Context, cluster client.ObjectKey) (*client.DelegatingClient, error) {
-	m.delegatingClientsLock.Lock()
-	defer m.delegatingClientsLock.Unlock()
+	t.log.V(2).Info("Deleting accessor", "cluster", cluster.String())
 
-	// If another goroutine created the client while this one was waiting to acquire the write lock, return that
-	// instead of overwriting it.
-	if delegatingClient, exists := m.delegatingClients[cluster]; exists {
-		return delegatingClient, nil
-	}
+	t.log.V(4).Info("Stopping cache", "cluster", cluster.String())
+	a.cache.Stop()
+	t.log.V(4).Info("Cache stopped", "cluster", cluster.String())
 
-	cache, err := m.getOrCreateClusterCache(ctx, cluster)
-	if err != nil {
-		return nil, err
-	}
-	config, err := RESTConfig(ctx, m.client, cluster)
-	if err != nil {
-		return nil, errors.Wrap(err, "error fetching REST client config for remote cluster")
-	}
-	config.Timeout = defaultClientTimeout
-	c, err := client.New(config, client.Options{Scheme: m.scheme, Mapper: cache.mapper})
-	if err != nil {
-		return nil, err
-	}
-	delegatingClient := &client.DelegatingClient{
-		Reader:       cache,
-		Writer:       c,
-		StatusClient: c,
-	}
-	m.delegatingClients[cluster] = delegatingClient
-	return delegatingClient, nil
+	delete(t.accessors, cluster)
 }
 
-func (m *ClusterCacheTracker) deleteDelegatingClient(cluster client.ObjectKey) {
-	m.delegatingClientsLock.Lock()
-	defer m.delegatingClientsLock.Unlock()
-
-	delete(m.delegatingClients, cluster)
+// Watcher is a scoped-down interface from Controller that only knows how to watch.
+type Watcher interface {
+	// Watch watches src for changes, sending events to eventHandler if they pass predicates.
+	Watch(src source.Source, eventHandler handler.EventHandler, predicates ...predicate.Predicate) error
 }
 
-// getOrCreateClusterCache returns the clusterCache for cluster, creating a new ClusterCache if needed.
-func (m *ClusterCacheTracker) getOrCreateClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
-	cache := m.getClusterCache(cluster)
-	if cache != nil {
-		return cache, nil
-	}
+// WatchInput specifies the parameters used to establish a new watch for a remote cluster.
+type WatchInput struct {
+	// Name represents a unique watch request.
```

Comment:
might be worth specify the uniqueness is referred to the set of watches for a given cluster